File: libs/statemanager/src/main/kotlin/mono/state/OneTimeActionHandler.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.state

import mono.actionmanager.OneTimeActionType
import mono.bitmap.manager.MonoBitmapManager
import mono.common.exhaustive
import mono.graphics.geo.Point
import mono.graphics.geo.Rect
import mono.lifecycle.LifecycleOwner
import mono.livedata.LiveData
import mono.shape.ShapeExtraManager
import mono.shape.clipboard.ShapeClipboardManager
import mono.shape.command.ChangeBound
import mono.shape.command.ChangeExtra
import mono.shape.command.ChangeOrder
import mono.shape.command.MakeTextEditable
import mono.shape.command.UpdateTextEditingMode
import mono.shape.extra.style.RectangleBorderCornerPattern
import mono.shape.extra.style.TextAlign
import mono.shape.shape.AbstractShape
import mono.shape.shape.Group
import mono.shape.shape.Line
import mono.shape.shape.MockShape
import mono.shape.shape.Rectangle
import mono.shape.shape.Text
import mono.state.command.CommandEnvironment
import mono.state.command.text.EditTextShapeHelper
import mono.state.onetimeaction.AppSettingActionHelper
import mono.state.onetimeaction.FileRelatedActionsHelper
import mono.state.utils.UpdateShapeBoundHelper
import mono.ui.appstate.AppUiStateManager

/**
 * A class to handle one time actions.
 */
internal class OneTimeActionHandler(
    lifecycleOwner: LifecycleOwner,
    oneTimeActionLiveData: LiveData<OneTimeActionType>,
    private val environment: CommandEnvironment,
    bitmapManager: MonoBitmapManager,
    shapeClipboardManager: ShapeClipboardManager,
    private val stateHistoryManager: StateHistoryManager,
    uiStateManager: AppUiStateManager
) {
    private val clipboardManager: ClipboardManager =
        ClipboardManager(lifecycleOwner, environment, shapeClipboardManager)

    private val fileRelatedActionsHelper = FileRelatedActionsHelper(
        environment,
        bitmapManager,
        shapeClipboardManager
    )

    private val appSettingActionHelper = AppSettingActionHelper(uiStateManager)

    init {
        oneTimeActionLiveData.observe(lifecycleOwner) {
            when (it) {
                OneTimeActionType.Idle -> Unit

                is OneTimeActionType.ProjectAction ->
                    fileRelatedActionsHelper.handleProjectAction(it)

                is OneTimeActionType.AppSettingAction ->
                    appSettingActionHelper.handleAppSettingAction(it)

                // ---------
                OneTimeActionType.CopyText ->
                    fileRelatedActionsHelper.exportSelectedShapes(false)
                // ---------
                OneTimeActionType.SelectAllShapes ->
                    environment.selectAllShapes()

                OneTimeActionType.DeselectShapes ->
                    environment.clearSelectedShapes()

                OneTimeActionType.DeleteSelectedShapes ->
                    deleteSelectedShapes()

                OneTimeActionType.EditSelectedShapes ->
                    editSelectedShape(environment.getSelectedShapes().singleOrNull())

                is OneTimeActionType.EditSelectedShape ->
                    editSelectedShape(it.shape)

                is OneTimeActionType.TextAlignment ->
                    setTextAlignment(it.newHorizontalAlign, it.newVerticalAlign)
                // ---------
                is OneTimeActionType.MoveShapes ->
                    moveSelectedShapes(it.offsetRow, it.offsetCol)

                is OneTimeActionType.ChangeShapeBound ->
                    setSelectedShapeBound(it.newLeft, it.newTop, it.newWidth, it.newHeight)
                // ---------
                is OneTimeActionType.ChangeShapeFillExtra ->
                    setSelectedShapeFillExtra(it.isEnabled, it.newFillStyleId)

                is OneTimeActionType.ChangeShapeBorderExtra ->
                    setSelectedShapeBorderExtra(it.isEnabled, it.newBorderStyleId)

                is OneTimeActionType.ChangeShapeBorderDashPatternExtra ->
                    setSelectedShapeBorderDashPatternExtra(it.dash, it.gap, it.offset)

                is OneTimeActionType.ChangeShapeBorderCornerExtra ->
                    setSelectedShapeBorderCornerExtra(it.isRoundedCorner)

                is OneTimeActionType.ChangeLineStrokeExtra ->
                    setSelectedLineStrokeExtra(it.isEnabled, it.newStrokeStyleId)

                is OneTimeActionType.ChangeLineStrokeDashPatternExtra ->
                    setSelectedLineStrokeDashPattern(it.dash, it.gap, it.offset)

                is OneTimeActionType.ChangeLineStrokeCornerExtra ->
                    setSelectedLineStrokeCornerExtra(it.isRoundedCorner)

                is OneTimeActionType.ChangeLineStartAnchorExtra ->
                    setSelectedShapeStartAnchorExtra(it.isEnabled, it.newHeadId)

                is OneTimeActionType.ChangeLineEndAnchorExtra ->
                    setSelectedShapeEndAnchorExtra(it.isEnabled, it.newHeadId)
                // ---------
                is OneTimeActionType.ReorderShape ->
                    changeShapeOrder(it.orderType)
                // ---------
                is OneTimeActionType.Copy ->
                    clipboardManager.copySelectedShapes(it.isRemoveRequired)

                OneTimeActionType.Duplicate ->
                    clipboardManager.duplicateSelectedShapes()
                // ---------
                OneTimeActionType.Undo ->
                    stateHistoryManager.undo()

                OneTimeActionType.Redo ->
                    stateHistoryManager.redo()
            }.exhaustive
        }
    }

    private fun deleteSelectedShapes() {
        for (shape in environment.getSelectedShapes()) {
            environment.removeShape(shape)
            environment.shapeManager.shapeConnector.removeShape(shape)
        }
        environment.clearSelectedShapes()
    }

    private fun editSelectedShape(shape: AbstractShape?) {
        when (shape) {
            is Text -> {
                environment.enterEditingMode()
                val oldText = shape.text
                environment.shapeManager.execute(MakeTextEditable(shape))
                environment.shapeManager.execute(UpdateTextEditingMode(shape, true))
                EditTextShapeHelper.showEditTextDialog(environment, shape, false) {
                    environment.shapeManager.execute(UpdateTextEditingMode(shape, false))
                    environment.exitEditingMode(oldText != shape.text)
                }
            }

            is Line,
            is Rectangle,
            is MockShape,
            is Group,
            null -> Unit
        }.exhaustive
    }

    private fun setTextAlignment(
        horizontalAlign: TextAlign.HorizontalAlign?,
        verticalAlign: TextAlign.VerticalAlign?
    ) {
        val textShape = environment.getSelectedShapes().singleOrNull() as? Text
        if (textShape == null) {
            ShapeExtraManager.setDefaultValues(
                textHorizontalAlign = horizontalAlign,
                textVerticalAlign = verticalAlign
            )
            return
        }
        val newTextAlign = textShape.extra.textAlign.copy(
            horizontalAlign ?: textShape.extra.textAlign.horizontalAlign,
            verticalAlign ?: textShape.extra.textAlign.verticalAlign
        )
        val newExtra = textShape.extra.copy(textAlign = newTextAlign)
        environment.shapeManager.execute(ChangeExtra(textShape, newExtra))
    }

    private fun moveSelectedShapes(offsetRow: Int, offsetCol: Int) {
        val offset = Point(offsetCol, offsetRow)
        UpdateShapeBoundHelper.moveShapes(
            environment,
            environment.getSelectedShapes(),
            isUpdateConfirmed = true
        ) { it.bound.position + offset }

        environment.updateInteractionBounds()
    }

    private fun setSelectedShapeBound(left: Int?, top: Int?, width: Int?, height: Int?) {
        val singleShape = environment.getSelectedShapes().singleOrNull() ?: return
        val currentBound = singleShape.bound
        val newBound = Rect.byLTWH(
            left = left ?: currentBound.left,
            top = top ?: currentBound.top,
            width = width ?: currentBound.width,
            height = height ?: currentBound.height
        )
        UpdateShapeBoundHelper.updateConnectors(environment, singleShape, newBound, true)
        environment.shapeManager.execute(
            ChangeBound(singleShape, newBound)
        )
        environment.updateInteractionBounds()
    }

    private fun setSelectedShapeFillExtra(isEnabled: Boolean?, newFillStyleId: String?) {
        val singleShape = environment.getSelectedShapes().singleOrNull()

        if (singleShape == null) {
            ShapeExtraManager.setDefaultValues(
                isFillEnabled = isEnabled,
                fillStyleId = newFillStyleId
            )
            return
        }

        val currentRectangleExtra = when (singleShape) {
            is Rectangle -> singleShape.extra
            is Text -> singleShape.extra.boundExtra
            is Line,
            is MockShape,
            is Group -> null
        } ?: return
        val newIsFillEnabled = isEnabled ?: currentRectangleExtra.isFillEnabled
        val newFillStyle = ShapeExtraManager.getRectangleFillStyle(
            newFillStyleId,
            currentRectangleExtra.userSelectedFillStyle
        )
        val rectangleExtra = currentRectangleExtra.copy(
            isFillEnabled = newIsFillEnabled,
            userSelectedFillStyle = newFillStyle
        )
        val newExtra = when (singleShape) {
            is Rectangle -> rectangleExtra
            is Text -> singleShape.extra.copy(boundExtra = rectangleExtra)
            is Line,
            is MockShape,
            is Group -> null
        } ?: return
        environment.shapeManager.execute(ChangeExtra(singleShape, newExtra))
    }

    private fun setSelectedShapeBorderExtra(isEnabled: Boolean?, newBorderStyleId: String?) {
        val singleShape = environment.getSelectedShapes().singleOrNull()
        if (singleShape == null) {
            ShapeExtraManager.setDefaultValues(
                isBorderEnabled = isEnabled,
                borderStyleId = newBorderStyleId
            )
            return
        }

        val currentRectangleExtra = when (singleShape) {
            is Rectangle -> singleShape.extra
            is Text -> singleShape.extra.boundExtra
            is Line,
            is MockShape,
            is Group -> null
        } ?: return
        val newIsBorderEnabled = isEnabled ?: currentRectangleExtra.isBorderEnabled
        val newBorderStyle = ShapeExtraManager.getRectangleBorderStyle(
            newBorderStyleId,
            currentRectangleExtra.userSelectedBorderStyle
        )
        val rectangleExtra = currentRectangleExtra.copy(
            isBorderEnabled = newIsBorderEnabled,
            userSelectedBorderStyle = newBorderStyle
        )
        val newExtra = when (singleShape) {
            is Rectangle -> rectangleExtra
            is Text -> singleShape.extra.copy(boundExtra = rectangleExtra)
            is Line,
            is MockShape,
            is Group -> null
        } ?: return
        environment.shapeManager.execute(ChangeExtra(singleShape, newExtra))
    }

    private fun setSelectedShapeBorderDashPatternExtra(dash: Int?, gap: Int?, offset: Int?) {
        val singleShape = environment.getSelectedShapes().singleOrNull()
        if (singleShape == null) {
            val currentDefaultDashPattern = ShapeExtraManager.defaultRectangleExtra.dashPattern
            val newDefaultDashPattern = currentDefaultDashPattern.copy(
                dash = dash ?: currentDefaultDashPattern.dash,
                gap = gap ?: currentDefaultDashPattern.gap,
                offset = offset ?: currentDefaultDashPattern.offset
            )
            ShapeExtraManager.setDefaultValues(borderDashPattern = newDefaultDashPattern)
            return
        }

        val rectangleExtra = when (singleShape) {
            is Rectangle -> singleShape.extra
            is Text -> singleShape.extra.boundExtra
            is Group,
            is Line,
            is MockShape -> null
        }
        val currentPattern = rectangleExtra?.dashPattern ?: return
        val newPattern = currentPattern.copy(
            dash = dash ?: currentPattern.dash,
            gap = gap ?: currentPattern.gap,
            offset = offset ?: currentPattern.offset
        )
        val newRectangleExtra = rectangleExtra.copy(dashPattern = newPattern)
        val newExtra = when (singleShape) {
            is Rectangle -> newRectangleExtra
            is Text -> singleShape.extra.copy(boundExtra = newRectangleExtra)
            is Group,
            is Line,
            is MockShape -> null
        } ?: return
        environment.shapeManager.execute(ChangeExtra(singleShape, newExtra))
    }

    private fun setSelectedShapeBorderCornerExtra(isRoundedCorner: Boolean) {
        val singleShape = environment.getSelectedShapes().singleOrNull()
        val rectangleExtra = when (singleShape) {
            is Rectangle -> singleShape.extra
            is Text -> singleShape.extra.boundExtra
            is Group,
            is Line,
            is MockShape,
            null -> null
        }
        if (singleShape == null || rectangleExtra == null) {
            ShapeExtraManager.setDefaultValues(isBorderRoundedCorner = isRoundedCorner)
            return
        }
        val newCorner =
            if (isRoundedCorner) {
                RectangleBorderCornerPattern.ENABLED
            } else {
                RectangleBorderCornerPattern.DISABLED
            }
        val newRectangleExtra = rectangleExtra.copy(corner = newCorner)
        val newExtra = when (singleShape) {
            is Rectangle -> newRectangleExtra
            is Text -> singleShape.extra.copy(boundExtra = newRectangleExtra)
            is Group,
            is Line,
            is MockShape -> null
        } ?: return
        environment.shapeManager.execute(ChangeExtra(singleShape, newExtra))
    }

    private fun setSelectedLineStrokeExtra(isEnabled: Boolean?, newStrokeStyleId: String?) {
        val line = environment.getSelectedShapes().singleOrNull() as? Line
        if (line == null) {
            ShapeExtraManager.setDefaultValues(
                isLineStrokeEnabled = isEnabled,
                lineStrokeStyleId = newStrokeStyleId
            )
            return
        }

        val currentLineExtra = line.extra
        val newIsEnabled = isEnabled ?: currentLineExtra.isStrokeEnabled
        val newStrokeStyle = ShapeExtraManager.getLineStrokeStyle(
            newStrokeStyleId,
            currentLineExtra.userSelectedStrokeStyle
        )
        val newExtra = currentLineExtra.copy(
            isStrokeEnabled = newIsEnabled,
            userSelectedStrokeStyle = newStrokeStyle
        )
        environment.shapeManager.execute(ChangeExtra(line, newExtra))
    }

    private fun setSelectedLineStrokeDashPattern(dash: Int?, gap: Int?, offset: Int?) {
        val line = environment.getSelectedShapes().singleOrNull() as? Line
        if (line == null) {
            val currentDefaultDashPattern = ShapeExtraManager.defaultLineExtra.dashPattern
            val newDefaultDashPattern = currentDefaultDashPattern.copy(
                dash = dash ?: currentDefaultDashPattern.dash,
                gap = gap ?: currentDefaultDashPattern.gap,
                offset = offset ?: currentDefaultDashPattern.offset
            )
            ShapeExtraManager.setDefaultValues(lineDashPattern = newDefaultDashPattern)
            return
        }
        val currentExtra = line.extra
        val currentPattern = currentExtra.dashPattern
        val newPattern = currentPattern.copy(
            dash = dash ?: currentPattern.dash,
            gap = gap ?: currentPattern.gap,
            offset = offset ?: currentPattern.offset
        )
        val newExtra = currentExtra.copy(dashPattern = newPattern)
        environment.shapeManager.execute(ChangeExtra(line, newExtra))
    }

    private fun setSelectedLineStrokeCornerExtra(isRoundedCorner: Boolean) {
        val line = environment.getSelectedShapes().singleOrNull() as? Line
        if (line == null) {
            ShapeExtraManager.setDefaultValues(isLineStrokeRoundedCorner = isRoundedCorner)
            return
        }
        val currentExtra = line.extra
        val newExtra = currentExtra.copy(isRoundedCorner = isRoundedCorner)

        environment.shapeManager.execute(ChangeExtra(line, newExtra))
    }

    private fun setSelectedShapeStartAnchorExtra(isEnabled: Boolean?, newAnchorId: String?) {
        val line = environment.getSelectedShapes().singleOrNull() as? Line
        if (line == null) {
            ShapeExtraManager.setDefaultValues(
                isStartHeadAnchorCharEnabled = isEnabled,
                startHeadAnchorCharId = newAnchorId
            )
            return
        }

        val currentExtra = line.extra
        val newIsEnabled = isEnabled ?: currentExtra.isStartAnchorEnabled
        val newAnchor =
            ShapeExtraManager.getStartHeadAnchorChar(
                newAnchorId,
                currentExtra.userSelectedStartAnchor
            )
        val newExtra = currentExtra.copy(
            isStartAnchorEnabled = newIsEnabled,
            userSelectedStartAnchor = newAnchor
        )
        environment.shapeManager.execute(ChangeExtra(line, newExtra))
    }

    private fun setSelectedShapeEndAnchorExtra(isEnabled: Boolean?, newAnchorId: String?) {
        val line = environment.getSelectedShapes().singleOrNull() as? Line
        if (line == null) {
            ShapeExtraManager.setDefaultValues(
                isEndHeadAnchorCharEnabled = isEnabled,
                endHeadAnchorCharId = newAnchorId
            )
            return
        }
        val currentExtra = line.extra
        val newIsEnabled = isEnabled ?: currentExtra.isEndAnchorEnabled
        val newAnchor =
            ShapeExtraManager.getEndHeadAnchorChar(
                newAnchorId,
                currentExtra.userSelectedEndAnchor
            )
        val newExtra = currentExtra.copy(
            isEndAnchorEnabled = newIsEnabled,
            userSelectedEndAnchor = newAnchor
        )
        environment.shapeManager.execute(ChangeExtra(line, newExtra))
    }

    private fun changeShapeOrder(orderType: ChangeOrder.ChangeOrderType) {
        val singleShape = environment.getSelectedShapes().singleOrNull() ?: return
        environment.shapeManager.execute(ChangeOrder(singleShape, orderType))
    }
}

File: monosketch-svelte/src/lib/mono/state-manager/one-time-action-handler.ts
/*
 * Copyright (c) 2024, tuanchauict
 */

import { Flow, LifecycleOwner } from "$libs/flow";
import { Point } from "$libs/graphics-geo/point";
import { Rect } from "$libs/graphics-geo/rect";
import { singleOrNull } from "$libs/sequence";
import type { OneTimeActionType } from "$mono/action-manager/one-time-actions";
import type { MonoBitmapManager } from "$mono/monobitmap/manager/mono-bitmap-manager";
import { ChangeBound, ChangeExtra } from "$mono/shape/command/general-shape-commands";
import { ChangeOrder, type ChangeOrderType } from "$mono/shape/command/shape-manager-commands";
import { MakeTextEditable, UpdateTextEditingMode } from "$mono/shape/command/text-commands";
import { ShapeExtraManager } from "$mono/shape/extra/extra-manager";
import { TextAlign, type TextHorizontalAlign, TextVerticalAlign } from "$mono/shape/extra/style";
import type { ShapeClipboardManager } from "$mono/shape/shape-clipboard-manager";
import type { AbstractShape } from "$mono/shape/shape/abstract-shape";
import type { Line } from "$mono/shape/shape/line";
import { Rectangle } from "$mono/shape/shape/rectangle";
import { Text } from "$mono/shape/shape/text";
import { ClipboardManager } from "$mono/state-manager/clipboard-manager";
import type { CommandEnvironment } from "$mono/state-manager/command-environment";
import { EditTextShapeHelper } from "$mono/state-manager/command/text/edit-text-shape-helper";
import { FileRelatedActionsHelper } from "$mono/state-manager/onetimeaction/file-related-action-helper";
import type { StateHistoryManager } from "$mono/state-manager/state-history-manager";
import { moveShapes, updateConnectors } from "$mono/state-manager/utils/UpdateShapeBoundHelper";
import type { WorkspaceDao } from "$mono/store-manager/dao/workspace-dao";
import type { AppUiStateManager } from "$mono/ui-state-manager/app-ui-state-manager";

/**
 * A class to handle one time actions.
 */
export class OneTimeActionHandler {
    private readonly fileRelatedActionsHelper: FileRelatedActionsHelper;
    private readonly clipboardManager: ClipboardManager;

    constructor(
        private readonly environment: CommandEnvironment,
        private readonly bitmapManager: MonoBitmapManager,
        private readonly shapeClipboardManager: ShapeClipboardManager,
        private readonly stateHistoryManager: StateHistoryManager,
        private readonly uiStateManger: AppUiStateManager,
        private readonly workspaceDao: WorkspaceDao,
    ) {
        this.fileRelatedActionsHelper = new FileRelatedActionsHelper(
            environment,
            bitmapManager,
            shapeClipboardManager,
            workspaceDao,
        );
        this.clipboardManager = new ClipboardManager(environment, shapeClipboardManager);
    }

    observe(lifecycleOwner: LifecycleOwner, oneTimeActionFlow: Flow<OneTimeActionType>) {
        oneTimeActionFlow.observe(lifecycleOwner, (action) => {
            switch (action.type) {
                case "Idle":
                    break;

                case "RenameCurrentProject":
                case "NewProject":
                case "ExportSelectedShapes":
                case "SwitchProject":
                case "RemoveProject":
                case "SaveShapesAs":
                case "OpenShapes":
                    this.fileRelatedActionsHelper.handleProjectAction(action);
                    break;

                // ---------
                case "CopyText":
                    this.fileRelatedActionsHelper.exportSelectedShapes(false);
                    break;
                // ---------
                case "SelectAllShapes":
                    this.environment.selectAllShapes();
                    break;

                case "DeselectShapes":
                    this.environment.clearSelectedShapes();
                    break;

                case "DeleteSelectedShapes":
                    this.deleteSelectedShapes();
                    break;

                case "EditSelectedShapes":
                    this.editSelectedShape(singleOrNull(this.environment.getSelectedShapes()));
                    break;

                case "EditSelectedShape":
                    this.editSelectedShape(action.shape);
                    break;

                case "TextAlignment":
                    this.setTextAlignment(action.newHorizontalAlign, action.newVerticalAlign);
                    break;
                // ---------
                case "MoveShapes":
                    this.moveSelectedShapes(action.offsetRow, action.offsetCol);
                    break;

                case "ChangeShapeBound":
                    this.setSelectedShapeBound(action.newLeft, action.newTop, action.newWidth, action.newHeight);
                    break;
                // ---------
                case "ChangeShapeFillExtra":
                    this.setSelectedShapeFillExtra(action.isEnabled, action.newFillStyleId);
                    break;

                case "ChangeShapeBorderExtra":
                    this.setSelectedShapeBorderExtra(action.isEnabled, action.newBorderStyleId);
                    break;

                case "ChangeShapeBorderDashPatternExtra":
                    this.setSelectedShapeBorderDashPatternExtra(action.dash, action.gap, action.offset);
                    break;

                case "ChangeShapeBorderCornerExtra":
                    this.setSelectedShapeBorderCornerExtra(action.isRoundedCorner);
                    break;

                case "ChangeLineStrokeExtra":
                    this.setSelectedLineStrokeExtra(action.isEnabled, action.newStrokeStyleId);
                    break;

                case "ChangeLineStrokeDashPatternExtra":
                    this.setSelectedLineStrokeDashPattern(action.dash, action.gap, action.offset);
                    break;

                case "ChangeLineStrokeCornerExtra":
                    this.setSelectedLineStrokeCornerExtra(action.isRoundedCorner);
                    break;

                case "ChangeLineStartAnchorExtra":
                    this.setSelectedShapeStartAnchorExtra(action.isEnabled, action.newHeadId);
                    break;

                case "ChangeLineEndAnchorExtra":
                    this.setSelectedShapeEndAnchorExtra(action.isEnabled, action.newHeadId);
                    break;
                // ---------
                case "ReorderShape":
                    this.changeShapeOrder(action.orderType);
                    break;
                // ---------
                case "Copy":
                    this.clipboardManager.copySelectedShapes(action.isRemoveRequired);
                    break;

                case "Duplicate":
                    this.clipboardManager.duplicateSelectedShapes();
                    break;
                // ---------
                case "Undo":
                    this.stateHistoryManager.undo();
                    break;

                case "Redo":
                    this.stateHistoryManager.redo();
                    break;
            }
        });
    }

    private deleteSelectedShapes() {
        const environment = this.environment;
        for (const shape of environment.getSelectedShapes()) {
            environment.removeShape(shape);
            // TODO: Add removeConnectorsOfShape() to CommandEnvironment
            environment.shapeManager.shapeConnector.removeShape(shape);
        }
        environment.clearSelectedShapes();
    }

    private editSelectedShape(shape: AbstractShape | null) {
        if (shape instanceof Text) {
            this.environment.enterEditingMode();
            const oldText = shape.text;
            this.environment.shapeManager.execute(new MakeTextEditable(shape));
            this.environment.shapeManager.execute(new UpdateTextEditingMode(shape, true));

            console.log("Edit text shape", oldText);
            EditTextShapeHelper.showEditTextDialog(this.environment, shape, false, () => {
                this.environment.shapeManager.execute(new UpdateTextEditingMode(shape, false));
                this.environment.exitEditingMode(oldText !== shape.text);
            });
        }
    }

    private setTextAlignment(newHorizontalAlign: TextHorizontalAlign, newVerticalAlign: TextVerticalAlign) {
        const textShape = this.singleSelectedShape() as Text | null;
        if (textShape === null) {
            ShapeExtraManager.setDefaultValues({
                textHorizontalAlign: newHorizontalAlign,
                textVerticalAlign: newVerticalAlign,
            });
            return;
        }
        const extra = textShape.extra;
        const newTextAlign = new TextAlign(
            newHorizontalAlign ?? extra.textAlign.horizontalAlign,
            newVerticalAlign ?? extra.textAlign.verticalAlign,
        );
        const newExtra = extra.copy({ textAlign: newTextAlign });
        this.environment.shapeManager.execute(new ChangeExtra(textShape, newExtra));
    }

    private moveSelectedShapes(offsetRow: number, offsetCol: number) {
        const offset = Point.of(offsetCol, offsetRow);
        moveShapes(
            this.environment,
            Array.from(this.environment.getSelectedShapes()),
            true,
            (shape) => shape.bound.position.plus(offset),
        );
        this.environment.updateInteractionBounds();
    }

    private setSelectedShapeBound(newLeft: number, newTop: number, newWidth: number, newHeight: number) {
        const singleShape = this.singleSelectedShape();
        if (singleShape === null) {
            return;
        }
        const newBound = Rect.byLTWH(
            newLeft ?? singleShape.bound.left,
            newTop ?? singleShape.bound.top,
            newWidth ?? singleShape.bound.width,
            newHeight ?? singleShape.bound.height,
        );
        updateConnectors(this.environment, singleShape, newBound, true);
        this.environment.shapeManager.execute(new ChangeBound(singleShape, newBound));
        this.environment.updateInteractionBounds();
    }

    private setSelectedShapeFillExtra(isEnabled: boolean, newFillStyleId: string | null) {
        const singleShape = this.singleSelectedShape();
        if (singleShape === null) {
            ShapeExtraManager.setDefaultValues({ isFillEnabled: isEnabled, fillStyleId: newFillStyleId ?? undefined });
            return;
        }

        const currentRectangleExtra = singleShape instanceof Text ? singleShape.extra.boundExtra :
            (singleShape instanceof Rectangle ? singleShape.extra : null);
        if (currentRectangleExtra === null) {
            return;
        }

        const newIsFillEnabled = isEnabled ?? currentRectangleExtra.isFillEnabled;
        const newFillStyle = ShapeExtraManager.getRectangleFillStyle(
            newFillStyleId ?? undefined,
            currentRectangleExtra.userSelectedFillStyle,
        );
        const rectangleExtra = currentRectangleExtra.copy({
            isFillEnabled: newIsFillEnabled,
            userSelectedFillStyle: newFillStyle,
        });
        const newExtra =
            singleShape instanceof Text ? singleShape.extra.copy({ boundExtra: rectangleExtra }) : rectangleExtra;
        this.environment.shapeManager.execute(new ChangeExtra(singleShape, newExtra));
    }

    private setSelectedShapeBorderExtra(isEnabled: boolean, newBorderStyleId: string | null) {

    }

    private setSelectedShapeBorderDashPatternExtra(dash: number, gap: number, offset: number) {

    }

    private setSelectedShapeBorderCornerExtra(isRoundedCorner: boolean) {

    }

    private setSelectedLineStrokeExtra(isEnabled: boolean, newStrokeStyleId: string | null) {

    }

    private setSelectedLineStrokeDashPattern(dash: number, gap: number, offset: number) {

    }

    private setSelectedLineStrokeCornerExtra(isRoundedCorner: boolean) {
        const line = this.singleSelectedShape() as Line | null;
        if (line === null) {
            ShapeExtraManager.setDefaultValues({ isLineStrokeRoundedCorner: isRoundedCorner });
            return;
        }
        const currentExtra = line.extra;
        const newExtra = currentExtra.copy({ isRoundedCorner: isRoundedCorner });
        this.environment.shapeManager.execute(new ChangeExtra(line, newExtra));
    }

    private setSelectedShapeStartAnchorExtra(isEnabled: boolean, newHeadId: string | null) {
        const line = this.singleSelectedShape() as Line | null;
        if (line === null) {
            ShapeExtraManager.setDefaultValues({
                isStartHeadAnchorCharEnabled: isEnabled,
                startHeadAnchorCharId: newHeadId ?? undefined,
            });
            return;
        }
        const currentExtra = line.extra;
        const newAnchor = ShapeExtraManager.getStartHeadAnchorChar(
            newHeadId ?? undefined,
            currentExtra.userSelectedEndAnchor,
        );
        const newExtra = currentExtra.copy({
            isStartAnchorEnabled: isEnabled ?? currentExtra.isEndAnchorEnabled,
            userSelectedStartAnchor: newAnchor,
        });
        this.environment.shapeManager.execute(new ChangeExtra(line, newExtra));
    }

    private setSelectedShapeEndAnchorExtra(isEnabled: boolean, newHeadId: string | null) {
        const line = this.singleSelectedShape() as Line | null;
        if (line === null) {
            ShapeExtraManager.setDefaultValues({
                isEndHeadAnchorCharEnabled: isEnabled,
                endHeadAnchorCharId: newHeadId ?? undefined,
            });
            return;
        }
        const currentExtra = line.extra;
        const newAnchor = ShapeExtraManager.getEndHeadAnchorChar(
            newHeadId ?? undefined,
            currentExtra.userSelectedEndAnchor,
        );
        const newExtra = currentExtra.copy({
            isEndAnchorEnabled: isEnabled ?? currentExtra.isEndAnchorEnabled,
            userSelectedEndAnchor: newAnchor,
        });
        this.environment.shapeManager.execute(new ChangeExtra(line, newExtra));
    }

    private changeShapeOrder(orderType: ChangeOrderType) {
        const singleSelectedShape = this.singleSelectedShape();
        if (singleSelectedShape !== null) {
            this.environment.shapeManager.execute(new ChangeOrder(singleSelectedShape, orderType));
        }
    }

    private singleSelectedShape(): AbstractShape | null {
        return singleOrNull(this.environment.getSelectedShapes());
    }
}
